### ISSUE TITLE
[BUG] Wrong local socket used after switching to docker-compose version 2 under MacOs

### ISSUE THREAD

**Issue by crashev** [docker/compose#10641]
### Description

After enabling docker-composer version 2 in Docker Desktop UI for MacOs, docker-compose started using `/var/run/docker.sock` instead of `~/.docker/run/docker.sock` which was used by docker composer version 1 and which should be used. 

It now generates the error: 
```
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon runnin
```

when running docker-compose up

### Steps To Reproduce

1. docker-compose enable-v2
2. docker-compose up

 vs
1. docker-compose disable-v2
2. docker-compose up

### Compose Version

```Text
$ docker compose version
Docker Compose version v2.18.1
$ docker-compose version
Docker Compose version v2.18.1

$ which docker-compose
/usr/local/bin/docker-compose
$ ls -ld /usr/local/bin/docker-compose
lrwxr-xr-x  1 root  wheel  62 Jul 18  2022 /usr/local/bin/docker-compose -> /Applications/Docker.app/Contents/Resources/bin/docker-compose
```

### Docker Environment

```Text
Client:
 Version:    24.0.2
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.5
    Path:     /Users/pfurtak/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.18.1
    Path:     /Users/pfurtak/.docker/cli-plugins/docker-compose
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     /Users/pfurtak/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.19
    Path:     /Users/pfurtak/.docker/cli-plugins/docker-extension
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v0.1.0-beta.4
    Path:     /Users/pfurtak/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/pfurtak/.docker/cli-plugins/docker-sbom
  scan: Docker Scan (Docker Inc.)
    Version:  v0.26.0
    Path:     /Users/pfurtak/.docker/cli-plugins/docker-scan
  scout: Command line tool for Docker Scout (Docker Inc.)
    Version:  v0.12.0
    Path:     /Users/pfurtak/.docker/cli-plugins/docker-scout

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 8
 Server Version: 24.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.49-linuxkit-pr
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 6
 Total Memory: 39.12GiB
 Name: docker-desktop
 ID: 655b01cb-2837-49e1-8e86-58990c711609
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by glours** [comment]
👋 @crashev 
Can you check in Docker Desktop settings that the `allow the default Docker socket to be used` option in the `Advanced` panel is checked?
![Screenshot 2023-06-01 at 10 26 25](https://github.com/docker/compose/assets/705411/148e4e64-a0ba-41ee-a28c-a05c4401c10e)

**Comment by crashev** [comment]
@glours Hey, yes, I know about this option, in my case, it's not enabled:

<img width="1855" alt="Zrzut ekranu 2023-06-1 o 10 37 15" src="https://github.com/docker/compose/assets/3481182/e2dda62e-ceb2-4f3d-8ebc-d1e4b6d5b4e0">

**Comment by glours** [comment]
So you got your explanation, you need this option to be activated to reach the `/var/run/docker.sock` or you modify your compose file to not rely on the socket directly (bind mounts...)
This is new security feature added to Docker Desktop, Compose v1 as a legacy version don't support this feature but Compose v2 does

**Comment by crashev** [comment]
If this option is not enabled shouldn't composer version 2 behave as version 1 and use local socket ? Or then maybe it should be documented that this option should be enabled if switching to docker compose version 2?

Can you elaborate more on "or you modify your compose file to not rely on the socket directly (bind mounts...)" as I didn't get your point here. In docker-compose file I can tell it to use local socket ?

**Comment by glours** [comment]
Compose v2 will use either `/var/run/docker.sock` or `~/.docker/run/docker.sock` depending on Docker Desktop configuration, you should not have to worry about this.
If you can share your compose file, we could see if there is an "hard" link to the `/var/run/docker.sock` which could explain your issue.
In some case people mount the Docker socket to their service containers to access to docker resources inside their containers

**Comment by crashev** [comment]
Ok I found the culprit for this. I realized that it's actually always failing on building

$ docker-compose up
[+] Building 0.0s (0/0)                                                                                                                                                                                         
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

The reason was the configuration of my builders:

```
$ docker buildx ls
NAME/NODE       DRIVER/ENDPOINT             STATUS  BUILDKIT                              PLATFORMS
mybuilder *     docker-container                                                          
  mybuilder0    unix:///var/run/docker.sock error                                         
default                                     error                                         
desktop-linux   docker                                                                    
  desktop-linux desktop-linux               running v0.11.7-0.20230525183624-798ad6b0ce9f linux/arm64, linux/amd64, linux/amd64/v2, linux/riscv64, linux/ppc64le, linux/s390x, linux/386, linux/mips64le, linux/mips64, linux/arm/v7, linux/arm/v6

Failed to get status for mybuilder * (mybuilder0): Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Cannot load builder default: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

Thank you @glours for your time on this.
Closing this bug ticket.
